include/pybind11/detail/type_caster_base.h

```diff
@@ -822,26 +822,192 @@ using movable_cast_op_type
                                   typename std::add_rvalue_reference<intrinsic_t<T>>::type,
                                   typename std::add_lvalue_reference<intrinsic_t<T>>::type>>;
 
-// std::is_copy_constructible isn't quite enough: it lets std::vector<T> (and similar) through when
-// T is non-copyable, but code containing such a copy constructor fails to actually compile.
-template <typename T, typename SFINAE = void>
-struct is_copy_constructible : std::is_copy_constructible<T> {};
+template <bool Condition, typename Then, typename Else>
+struct if_then_else {};
 
-template <typename T, typename SFINAE = void>
-struct is_move_constructible : std::is_move_constructible<T> {};
+template <typename Then, typename Else>
+struct if_then_else<true, Then, Else> {
+    using type = Then;
+};
+
+template <typename Then, typename Else>
+struct if_then_else<false, Then, Else> {
+    using type = Else;
+};
```

Comment:
Looks like you reinvented [std::conditional](https://en.cppreference.com/w/cpp/types/conditional).

Also, we have `conditional_t` in pybind11/detail/common.h.

I like `if_then_else` as a name a lot better than `conditional` btw, but standard is standard :-)